London | 25-SDC-July | Mikiyas Gebremichael | Sprint 1 | New Feature - Rebloom

Mikiyas-STP opened the pull request:
<!--

You must title your PR like this:

Region | Cohort | FirstName LastName | Sprint | Assignment Title

For example,

London | 25-ITP-May | Carol Owen | Sprint 1 | Alarm Clock

Fill in the template below - remove any sections that don't apply.

Complete the self checklist - replace each empty box in the checklist [ ] with a [x].

Add the label "Needs Review" and you will get review.

Respond to volunteer reviews until the volunteer marks it as "Complete".

-->

## Learners, PR Template

Self checklist

- [x] I have titled my PR with Region | Cohort | FirstName LastName | Sprint | Assignment Title
- [x] My changes meet the requirements of the task
- [x] I have tested my changes
- [x] My changes follow the [style guide](https://curriculum.codeyourfuture.io/guides/reviewing/style-guide/)

## Changelist

This PR is made for the rebloom implementation

## Questions

No question

Mikiyas-STP committed:
version of psycopg2 changed

Mikiyas-STP committed:
db : row added for the rebloom in schema

Mikiyas-STP committed:
dynamic input refactored on index.html

Mikiyas-STP committed:
Endpoints defined : Update rebloom counter and send rebloom

Mikiyas-STP committed:
Routing: rebloom

Mikiyas-STP committed:
SQL logic for bloom,rebloom and fetching data

Mikiyas-STP committed:
UI: rebloom button

Mikiyas-STP committed:
api endpoint for postrebloom and fetch bloom data ,

Mikiyas-STP committed:
Rebloom

illicitonion commented on backend/data/blooms.py:
```diff
@@ -121,18 +137,44 @@ def get_blooms_with_hashtag(
+        return None
+    content = original_bloom.content
+    update_rebloom_counter(id)
+    add_bloom(sender=sender, content=content, original_bloom_id=id)
```
Does it make sense setting the content for a rebloom?

Mikiyas-STP replied:
yes, because currently i copy the original content to let the rebloom appear as separate post(with its own timestamp) and link to the original bloom. if i change the way i do it i thought it would make the tracking rebloom time and sorting a bit complicated.

illicitonion commented on db/schema.sql:
```diff
@@ -10,7 +10,9 @@ CREATE TABLE blooms (
     content TEXT NOT NULL,
-    send_timestamp TIMESTAMP NOT NULL
+    send_timestamp TIMESTAMP NOT NULL,
+    reblooms INT NOT NULL DEFAULT 0,
```
Ideally you wouldn't need to track rebloom counts in the database - there is a race condition here where values can get out of date (e.g. if you create a rebloom but haven't updated this count yet).

Instead, use SQL relationships to compute this - in a query, you should be able to compute the count of blooms which have this ID as an original_bloom_id

Your Python and JS code for querying a Bloom shouldn't have to change, but it removes the need for you to update the original bloom from the python side.

Mikiyas-STP replied:
Thankyou

illicitonion replied:
This doesn't look like it changed?

Mikiyas-STP replied:
[rebloom count removed from the db to avoid race condition](https://github.com/CodeYourFuture/Module-Legacy-Code/pull/62/commits/cfbe7f9fd79cb9bcf9ce1a90f1b473a96f2e0768)

illicitonion commented on front-end/components/bloom.mjs:
```diff
@@ -30,6 +38,23 @@ const createBloom = (template, bloom) => {
+      .then((originalBloom) => {
+        const timeStamp = _formatTimestamp(originalBloom.sent_timestamp);
+        //I used inner html to render the arrow ↪ sign
+        rebloomInfoEl.innerHTML = `&#8618; Rebloom of the ${originalBloom.sender}'s post, posted ${timeStamp} ago`;
```
This is an XSS vulnerability - can you work out how to avoid that?

Mikiyas-STP replied:
Solved 

i have used createTextNode to prevent script excecution

Mikiyas-STP replied:
[60a4d01](https://github.com/CodeYourFuture/Module-Legacy-Code/pull/62/commits/60a4d01c7aae03707223f89d3189335493ee471c)

Mikiyas-STP replied:
[60a4d01](https://github.com/CodeYourFuture/Module-Legacy-Code/pull/62/commits/60a4d01c7aae03707223f89d3189335493ee471c) contain the issues related to rebloom and the small change for the issue with XSS Vulnerability

illicitonion reviewed (commented):
This generally looks good. A couple of UI considerations:
* Right now if I go to someone's profile page and press "Share", it reblooms, but the existing bloom doesn't update. What do you think of this UX?
* What sort order do you think reblooms should have? Should they sort at the original bloom time? Or the rebloom time? Or something different?

Mikiyas-STP committed:
Show rebloom as a new post with current timestamp and sort by DESC

Mikiyas-STP commented:
> This generally looks good. A couple of UI considerations:
> 
> * Right now if I go to someone's profile page and press "Share", it reblooms, but the existing bloom doesn't update. What do you think of this UX?
> * What sort order do you think reblooms should have? Should they sort at the original bloom time? Or the rebloom time? Or something different?

Hi @illicitonion Thank-you for your feedback. Here are the additional improvements i did. 
Now the existing bloom shows how many times it have been reblommed And i also used id DESC to display the most recent rebloom activity in the first place.

and as each rebloom creates a new bloom row with its timestamp the timeline shows the rebloom as a separate post this preserves my original bloom while the rebloom is visible.

Mikiyas-STP committed:
Fix: avoiding XSS Vulnerability

Mikiyas-STP committed:
rebloom count removed from the db to avoid race condition

illicitonion commented on backend/data/blooms.py:
```diff
@@ -13,21 +11,28 @@ class Bloom:
     with db_cursor() as cur:
         cur.execute(
-            "INSERT INTO blooms (id, sender_id, content, send_timestamp) VALUES (%(bloom_id)s, %(sender_id)s, %(content)s, %(timestamp)s)",
+            "INSERT INTO blooms (id, sender_id, content, send_timestamp, reblooms, original_bloom_id) VALUES (%(bloom_id)s, %(sender_id)s, %(content)s, %(timestamp)s, %(reblooms)s,%(original_bloom_id)s)",
```
Your INSERT here still writes to the column `reblooms` even though it no longer exists in your schema. This will always fail.

Mikiyas-STP replied:
Thankyou @illicitonion  I have carefully looked at it and removed all rebloom related implementation which doesn't match my SQL.
Please see the attached 3 commits for the changes , Let me know if i have other things to look into. i think this is my last PR to work on , Thanks

[dba2f49](https://github.com/CodeYourFuture/Module-Legacy-Code/pull/62/commits/dba2f49d19e7b893b3483d6bd2cb34cab4fc585d)

[a99c87e](https://github.com/CodeYourFuture/Module-Legacy-Code/pull/62/commits/a99c87e0504e0befb6f4891320a6dbe876db08be)

[c3fa67f](https://github.com/CodeYourFuture/Module-Legacy-Code/pull/62/commits/c3fa67fb07dc460146fba39bfb32d2bce3fddf0d)

Mikiyas-STP committed:
Fix: duplicate blooms in home timeline endpoint
Prevent selffollow duplication and
ensure blooms are not returned twice after refresh

Mikiyas-STP committed:
Update rebloom handling:
Display dynamic rebloom count from backend
and remove dependency on stored rebloom counter

Mikiyas-STP committed:
Fix: Remove stored rebloom counter and compute count via SQL

illicitonion commented:
LGTM, well done!

UI-wise, you may want to give more feedback when someone successfully reblooms, but I'll leave you to consider that if you care, I won't require it!

Mikiyas-STP commented:
> LGTM, well done!
> 
> UI-wise, you may want to give more feedback when someone successfully reblooms, but I'll leave you to consider that if you care, I won't require it!

Thank-You @illicitonion
